### ISSUE TITLE
Hash value verification issue image file meta data

### ISSUE THREAD

**Issue by malateshk** [Legrandin/pycryptodome#463]
I have an image file and generated a signature by hashing it and by using private key. I can successfully verify the hash of the image with public key and signature value. 
1.  hash(imagefile) + privatekey -> Sigfile
     I get successful verification when the same is verified with public key and signature.
2.  hash(imagefile) + publickey + Sigfile -> Successful
    I modify the meta data of the image file and i want the verification to fail since the original image file is being modified.
3.  hash(imagefile_metadata_change) publickey + Sigfile -> It should "Fail" but I get "Successful"

Python code:
```
     with open(public_key_filename) as f:
        public_key = RSA.importKey(f.read())

    data2 = base64.b64decode(data.get('imageFile').get('dataContent'))
    with open("test.jpg", 'wb') as f:
        f.write(data2)

    payload = SHA256.new(data2)
    print(payload)
    with open(signature_filename, 'rb') as f:
        signature = base64.b64decode(f.read())

    print(signature.hex())

    try:

        cipher = PKCS1_v1_5.new(public_key)
        cipher.verify(payload, signature)
        print("Success")
    except Exception as e:
        print(e)
        print('ERROR: signature verification failed')
```

Basically I am expecting the below command simulation in python
1. openssl dgst -sha256 -binary -out hash.bin image.jpg
2. openssl pkeyutl -inkey private.key -in hash.bin -out image.sig -sign
3. openssl pkeyutl -pubin -inkey public.key -in hash.bin -sigfile image.sig -verify

I couldn't figure out where the issue is !!

**Comment by texadactyl** [comment]
Please
1. Provide the entire source which includes all variable definitions and the imports that you used.
2. Provide all of the files that you used.

**Comment by malateshk** [comment]
Hi, attached the requested files. Below is the little background about my issue(linux code).
1. openssl dgst -sha256 -binary -out hash.bin
2. openssl dgst -sha256 -binary -out hash.bin download.jpeg
3. openssl pkeyutl -inkey private.key -in hash.bin -out data.sig -sign
4. openssl pkeyutl -pubin -inkey public.key -in hash.bin -sigfile data.sig -verify
    Signature Verified Successfully
--------------------------------------------------------------------------------------------------------------
5. I modified the metadata of download.jpeg -> download1.jpg
6. openssl dgst -sha256 -binary -out hash1.bin download.jpeg
7. openssl pkeyutl -pubin -inkey public.key -in hash1.bin -sigfile data.sig -verify
    Signature Verification Failure

My python code returns "Signature Verified Successfully" in the later case.
[Crypto.zip](https://github.com/Legrandin/pycryptodome/files/5487103/Crypto.zip) attached code and relevant files.

**Comment by texadactyl** [comment]
```
import cryptography.exceptions
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.primitives.serialization import load_pem_public_key
```

Your source code refers to the pypi package, "cryptography".
Did you file an issue in the right github site?

**Comment by malateshk** [comment]
@texadactyl, below is the import which I tried too
```
import base64

from Crypto.Signature import PKCS1_v1_5
from Crypto.PublicKey import RSA
from Crypto.Hash import SHA256


file_to_verify = "download1.jpg"
signature_filename = "data.sig"
public_key_filename = "public.key"
private_key_filename = "private.key"

def load_creative(file_path):
  with open(file_path, "rb") as image_file:
    return base64.b64encode(image_file.read())

def verify_signature():
    data = {'imageFile':{'fileName':'download1.jpg', 'dataContent':load_creative(file_to_verify).decode()},
            'signatureFile':{'filename':'data.sig', 'dataContent': load_creative(signature_filename).decode()}}

    with open(public_key_filename) as f:
        public_key = RSA.importKey(f.read())

    data2 = base64.b64decode(data.get('imageFile').get('dataContent'))

    h = SHA256.new()
    h.update(data2)
    payload = h.hexdigest()
    with open(signature_filename, 'rb') as f:
        signature = base64.b64decode(f.read())

    try:
        cipher = PKCS1_v1_5.new(public_key)
        cipher.verify(payload, signature)
        print("Success")
    except Exception as e:
        print(e)
        print('ERROR: signature verification failed')

verify_signature()

```

**Comment by texadactyl** [comment]
Using your supplied source and the files in the ZIP, I had to rename the JPEG to JPG.
Then, I ran the program and this happened:
```
    with open(signature_filename, 'rb') as f:
        signature = base64.b64decode(f.read())  ##  <------

  File "/mnt/elkdata/linux-home-folder/projects/crypto/issue_463/validate.py", line 30, in verify_signature
    signature = base64.b64decode(f.read())

  File "/usr/lib/python3.8/base64.py", line 87, in b64decode
    return binascii.a2b_base64(s)

Error: Incorrect padding
```

Summary:
- Input files (public key, private key, signature) were supplied without showing the code for creating them.
- Some confusing superfluous files (E.g. download.jpeg, hash.bin) were supplied.
- Code for creating input files is missing.  I see your process design but that is insufficient evidence.

**Comment by texadactyl** [comment]
How did you do this?
"I modified the metadata of download.jpeg -> download1.jpg"

**Comment by texadactyl** [comment]
6. openssl dgst -sha256 -binary -out hash1.bin download1.jpg

right?

**Comment by texadactyl** [comment]
When I run validate.py, I get this result:
```
Traceback (most recent call last):

  File "/mnt/elkdata/linux-home-folder/projects/crypto/issue_463/validate.py", line 40, in <module>
    verify_signature()

  File "/mnt/elkdata/linux-home-folder/projects/crypto/issue_463/validate.py", line 30, in verify_signature
    signature = base64.b64decode(f.read())

  File "/usr/lib/python3.8/base64.py", line 87, in b64decode
    return binascii.a2b_base64(s)

Error: Incorrect padding
```
So, the Python code does not successfully verify and it tells you why.

**Comment by texadactyl** [comment]
Do this instead:  signature = f.read()

Then, your program also fails to verify the signature:
"ERROR: signature verification failed"

Satisfied?  If so, please close this issue.

**Comment by texadactyl** [comment]
There is some confusion between string and bytes variables in your program.  I suggest you use an IDE (spyder or pycharm) or pylint command line to sort it out.

**Comment by texadactyl** [comment]
Why the JSON definition for "data".  It makes your test program unnecessarily hard to read.
